src/GitVersionCore/Model/VersionVariables.cs

```diff
@@ -199,19 +197,36 @@ private static bool ContainsKey(string variable)
 
         public override string ToString()
         {
-            var variables = this.GetProperties().ToDictionary(x => x.Key, x => x.Value);
+            var variablesType = typeof(VersionVariablesJsonModel);
+            var variables = new VersionVariablesJsonModel();
+
+            foreach (KeyValuePair<string, string> property in this.GetProperties())
+            {
+                variablesType.GetProperty(property.Key).SetValue(variables, property.Value);
+            }
+
             var serializeOptions = JsonSerializerOptions();
 
             return JsonSerializer.Serialize(variables, serializeOptions);
         }
 
+        private static JsonSerializerOptions JsonDeSerializerOptions()
+        {
+            var serializeOptions = new JsonSerializerOptions
+            {
+                WriteIndented = true,
+                Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
+                Converters = { new GitVersionNumberConverter() }
```

Comment:
With `GitVersionNumberConverter` explicitly added to all properties in `VersionVariablesJsonModel`, is it necessary to add to the `JsonSerializerOptions` as well?

Reply:
Ah, yes indeed. The property attributes take precedence over converters added to the `JsonSerializerOptions`. I've restored `JsonSerializerOptions `to its original state with a default `GitVersionStringConverter` configured. And removed the `JsonDeserializerOptions` because it is no longer needed.

The default `GitVersionStringConverter` is used when deserializing a json file where fields can be Json string or Json number but everything needs to be converted back into a `IEnumerable<KeyValuePair<string, string>>`.

`VersionVariablesJsonModel()` and `GitVersionNumberConverter()` are only used for serializing, so in that case the property attributes ensure Major, Minor, Patch, PreReleaseNumber, WeightedPreReleaseNumber, CommitsSinceVersionSource and UncommittedChanges are always serialized as JSON number.

It would also be possible to serialize null for these properties as the number 0 instead of Json null. That way the change in output would be kept to a minimum. Where it originally was:
```
  "PreReleaseNumber": "",
  "WeightedPreReleaseNumber": 0,
```

and the current proposal results in:
```
  "PreReleaseNumber": null,
  "WeightedPreReleaseNumber": 0,
```

this could also be considered:
and the current proposal has:
```
  "PreReleaseNumber": 0,
  "WeightedPreReleaseNumber": 0,
```

Note, I've renamed the extracted classes to have a VersionVariables prefix just so it's a bit more obvious they belong together.